src/proj6BittingCerratoCohenEllmer/controllers/CodeTabController.java

```diff
@@ -30,6 +35,50 @@
     private final HashMap<Tab,String> savedPaths = new HashMap<>();
 
     private final DialogHelper dialogHelper = new DialogHelper();
+    
+    // Vim Command Parameters
+    private boolean inVIMCommandMode = false;
+
+    //An Arraylist to keep track of commands for VIM chaining.
+    private final ArrayList<String> vimCommands = new ArrayList<String>();
+
+    private final EventHandler<KeyEvent> vimHandler = new EventHandler<KeyEvent>() {
+        /**
+         * Route VIM key press events to implemented methods
+         * 
+         * @param key A KeyEvent object that gives information about a Key Stroke
+         */
+        @Override
+        public void handle(KeyEvent key) {
+            // Escape -> command mode
+            if (KeyCode.ESCAPE.equals(key.getCode())) {
+                inVIMCommandMode = true;
+            }
+            // dont parse "command mode" command while in command mode!
+            if(inVIMCommandMode && !KeyCode.ESCAPE.equals(key.getCode()) ){
+                // must handle all key events to supress key stroke in codeArea
+                // Only read key released to prevent duplicated commands
+                if(key.getEventType().equals(KeyEvent.KEY_RELEASED)){
+                    //store commands for chaining
+                    vimCommands.add(key.getText());
+                    System.out.println("Vim Commands: " + vimCommands.toString());
+                    // 'i','I','a','A'-> edit mode
+                    if("i".equals(key.getText()) || "I".equals(key.getText()) ||
+                            "a".equals(key.getText()) || "A".equals(key.getText()) ){
+                        inVIMCommandMode = false;
```

Comment:
If in command mode, I type command+q, the window does not quit. I think this is the same bug as the red x button, but we'll probably have to create a workaround.